### ISSUE TITLE
New rule proposal: Literal Suffixes should be uppercase

### ISSUE THREAD

**Issue by berkeleybross** [DotNetAnalyzers/StyleCopAnalyzers#1563]
New rule proposal: Literal Suffixes should be uppercase

*Category*: Readability
*Rule*: Literal suffixes should be uppercase
*Rationale*: Improves readability of code, following recommendation from C# specification

Bad:
``` CSharp
long foo = 50l;
```

Good:
```CSharp
long foo = 50L;
```

See https://msdn.microsoft.com/en-us/library/aa664674(v=vs.71).aspx for the specification for integer literals.

**Comment by vweijsters** [comment]
:+1:

**Comment by Maxwe11** [comment]
I'd suggest restrict this only for L/UL suffixes

**Comment by sharwell** [comment]
I'm currently a "weak no" on this for the following reasons:

1. While it's clear to me that `L` is preferable to `l`, it's not so clear that this makes any difference for other suffixes. In fact, for the suffix `m` and the `e` which appears in floating point values, I almost feel like the lowercase version is more easily distinguished and possibly even more common in existing code.
2. I've never seen anyone use the lowercase suffix for integer literals.

@berkeleybross Have you looked for any instances of this syntax in open source projects?

**Comment by GregReddick** [comment]
I looked through my code, and I was mostly consistent on using lower case for m and f, and upper for L. I would be okay if the rule was consistent with lower for everything except L and possibly UL. It could also be configurable, but that really isn't in the StyleCop nature, which is provide one way of doing things. I did use L on integer constants, particularly when there were other constants that also has suffixes nearby. Example:
```
		private const long FileTimeOffset = 504911232000000000L;
		private const decimal MaxFileTime = 2650467743999999999m;
```
The fact that I was only mostly consistent might argue for having this rule.

**Comment by sharwell** [comment]
I'm now more of a :-1: on this because of [CS0078](https://msdn.microsoft.com/en-us/library/s74dtt7k.aspx).

**Comment by GregReddick** [comment]
I think there should be a consistent rule on the literal suffixes. I really don't care what it is, I can live with whatever is decided, but there should be one. It can't violate CS0078. I hate looking through my code and finding that I do something one way in one place and a different way in another place.

**Comment by otac0n** [comment]
Since this isn't closed, I'd cast my vote as :+1:.

**Comment by light-traveller** [comment]
Not necessarily. This is subjective. One reason can be readability as `m` is more readable than `M`; so the current situation with the existing compiler warning [CS0078](https://msdn.microsoft.com/en-us/library/s74dtt7k.aspx) are desirable. I am inclined to a negative vote.

> I think there should be a consistent rule on the literal suffixes. I really don't care what it is, I can live with whatever is decided, but there should be one. It can't violate CS0078. I hate looking through my code and finding that I do something one way in one place and a different way in another place.
